tracer/test/Datadog.Trace.Tests/SpanContextPropagatorTests.cs

```diff
@@ -3,34 +3,254 @@
+        {
+            var context = new SpanContext(TraceId, SpanId, SamplingPriority, serviceName: null, Origin);
+            var headers = new Mock<IHeadersCollection>();
 
-            result.TraceId.Should().Be(expectedTraceId);
-            result.SpanId.Should().Be(expectedSpanId);
-            result.SamplingPriority.Should().Be(expectedSamplingPriority);
-            result.Origin.Should().Be(expectedOrigin);
+            SpanContextPropagator.Instance.Inject(context, headers.Object);
+
+            VerifySetCalls(headers);
+        }
+
+        [Fact]
+        public void Inject_CarrierAndDelegate()
+        {
+            var context = new SpanContext(TraceId, SpanId, SamplingPriority, serviceName: null, Origin);
+            var headers = new Mock<IHeadersCollection>();
+
+            SpanContextPropagator.Instance.Inject(context, headers.Object, (carrier, name, value) => carrier.Set(name, value));
+
+            VerifySetCalls(headers);
+        }
+
+        [Fact]
+        public void Extract_IHeadersCollection()
+        {
+            // use `object` so Should() below works correctly,
+            // otherwise it picks up the IEnumerable<KeyValuePair<string, string>> overload
+            var headers = CreatePopulatedHeaders();
+            object result = SpanContextPropagator.Instance.Extract(headers.Object);
+
+            VerifyGetCalls(headers);
+
+            result.Should()
+                  .BeEquivalentTo(
+                       new
+                       {
+                           TraceId,
+                           SpanId,
+                           Origin,
+                           SamplingPriority
+                       });
+        }
+
+        [Fact]
+        public void Extract_CarrierAndDelegate()
+        {
+            // use `object` so Should() below works correctly,
+            // otherwise it picks up the IEnumerable<KeyValuePair<string, string>> overload
+            var headers = CreatePopulatedHeaders();
+            object result = SpanContextPropagator.Instance.Extract(headers.Object, (carrier, name) => carrier.GetValues(name));
+
+            VerifyGetCalls(headers);
+
+            result.Should()
+                  .BeEquivalentTo(
+                       new
+                       {
+                           TraceId,
+                           SpanId,
+                           Origin,
+                           SamplingPriority
+                       });
+        }
+
+        [Fact]
+        public void Extract_EmptyHeadersReturnsNull()
+        {
+            var headers = new Mock<IHeadersCollection>();
+
+            // use `object` so Should() below works correctly,
+            // otherwise it picks up the IEnumerable<KeyValuePair<string, string>> overload
+            object resultContext = SpanContextPropagator.Instance.Extract(headers.Object);
+
+            resultContext.Should().BeNull();
+        }
+
+        [Fact]
+        public void Identity()
```

Comment:
The test above this line are new. The tests below were moved here from `HeadersCollectionTestHelpers.cs` and refactored.